File: AgoraChat/Code/Controllers/Chats/AgoraChatConvUserDataModel.m
//
//  AgiraChatConvUserDataModel.m
//  EaseIM
//
//  Created by zhangchong on 2020/12/6.
//  Copyright © 2020 zhangchong. All rights reserved.
//

#import "AgoraChatConvUserDataModel.h"

@implementation AgoraChatConvUserDataModel

- (instancetype)initWithUserInfo:(AgoraChatUserInfo *)userInfo conversationType:(AgoraChatConversationType)type
{
    if (self = [super init]) {
        _easeId = userInfo.userId;
        _showName = userInfo.nickname;
        _avatarURL = userInfo.avatarUrl;
        _defaultAvatar = nil;
        
        if(type == AgoraChatConversationTypeGroupChat) {
            AgoraChatGroup* group = [AgoraChatGroup groupWithId:userInfo.userId];
            _showName = [group groupName];
            
            _defaultAvatar = ImageWithName(@"group_default_avatar");
        }
        
        if (type == AgoraChatConversationTypeChat) {
            UIImage *originImage = nil;
            
            NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
            NSString *imageName = [userDefault objectForKey:userInfo.userId];
            if (imageName && imageName.length > 0) {
                originImage = ImageWithName(imageName);
            } else {
                int random = arc4random() % 7 + 1;
                NSString *imgName = [NSString stringWithFormat:@"defatult_avatar_%@",@(random)];
                [userDefault setObject:imgName forKey:userInfo.userId];
                originImage = ImageWithName(imgName);
                [userDefault synchronize];
            }
            
            _defaultAvatar = [originImage acd_scaleToAssignSize:CGSizeMake(kAvatarHeight, kAvatarHeight)];
        }
    }
    return self;
}

- (UIImage*)_getDefaultAvatarImage:(NSString*)easeId conversationType:(AgoraChatConversationType)type
{
    if (type == AgoraChatConversationTypeChat) {
        return [UIImage imageNamed:@"defaultAvatar"];
    }
    if (type == AgoraChatConversationTypeGroupChat) {
        return [UIImage imageNamed:@"groupConversation"];
    }
    if (type == AgoraChatConversationTypeChatRoom) {
        return [UIImage imageNamed:@"chatroomConversation"];
    }
    return [UIImage imageNamed:@"defaultAvatar"];
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Chats/AgoraChatConvUserDataModel.swift
//
//  AgoraChatConvUserDataModel.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/17.
//

import UIKit

@objcMembers
class AgoraChatConvUserDataModel: NSObject, EaseUserProfile {
    
    var easeId: String {
        return _easeId ?? ""
    }
    
    var defaultAvatar: UIImage {
        return _defaultAvatar ?? UIImage()
    }
    
    var showName: String {
        return _showName ?? ""
    }
    
    var avatarURL: String {
        return _avatarURL ?? ""
    }
    
    let _easeId: String?
    var _defaultAvatar: UIImage?
    var _showName: String?
    let _avatarURL: String?
    
    init(userInfo: AgoraChatUserInfo, conversationType: AgoraChatConversationType) {
        self._easeId = userInfo.userId
        self._showName = userInfo.showName
        self._avatarURL = userInfo.avatarUrl
        self._defaultAvatar = nil
        if conversationType == .groupChat {
            let group = AgoraChatGroup(id: userInfo.userId)
            self._showName = group?.groupName
            self._defaultAvatar = UIImage(named: "group_default_avatar")
        } else if conversationType == .chat {
            let imageName = UserDefaults.standard.object(forKey: userInfo.userId ?? "") as? String
            var originImage: UIImage?
            if let imageName = imageName, imageName.count > 0 {
                originImage = UIImage(named: imageName)
            } else {
                let random = arc4random() % 7 + 1
                let imageName = "defatult_avatar_\(random)"
                originImage = UIImage(named: imageName)
                if let userId = userInfo.userId {
                    UserDefaults.standard.set(imageName, forKey: userId)
                    UserDefaults.standard.synchronize()
                }
            }
            self._defaultAvatar = originImage
        }
        super.init()
    }
}
